streaming/base/dataset/streaming.py

```diff
@@ -0,0 +1,638 @@
+            partitioned over the workers. Defaults to ``None``.
+        num_workers (int, optional): Number of workers of its DataLoader, which determines the size
+            of the barrier to coordinate workers while iterating. Defaults to ``None``.
+    """
+
+    def __init__(self,
+                 local: str,
+                 remote: Optional[str] = None,
+                 split: Optional[str] = None,
+                 shuffle: bool = False,
+                 predownload: Optional[int] = 100_000,
+                 keep_zip: Optional[bool] = None,
+                 download_retry: int = 2,
+                 download_timeout: float = 60,
+                 validate_hash: Optional[str] = None,
+                 seed: Optional[int] = None,
+                 batch_size: Optional[int] = None,
+                 num_workers: Optional[int] = None):
+        self.local = local
+        self.remote = remote
+        self.split = split or ''  # Empty string for os.path.join().
+        self.shuffle = shuffle
+        self.predownload = predownload
+        self.keep_zip = keep_zip
+        self.download_retry = download_retry
+        self.download_timeout = download_timeout
+        self.validate_hash = validate_hash or None
+        # Seed is set below.
+        self.batch_size = batch_size
+        self.num_workers = num_workers
+
+        # Load the index.json file.
+        basename = get_index_basename()
+        world = World()
+        if world.is_local_leader:
+            filename = self._download_file(basename)
+        else:
+            filename = os.path.join(local, split, basename)  # pyright: ignore
+        dist.barrier()
+        obj = json.load(open(filename))
+        assert obj['version'] == 2
+
+        # Initialize shard readers according to the loaded info.
+        self.shards = []
+        for info in obj['shards']:
+            shard = reader_from_json(local, split, info)
+            self.shards.append(shard)
+
+        # Build the Index (for partitioning and mapping samples to shards).
+        self.shard_sizes = np.array([x.samples for x in self.shards])
+        self.index = Index(self.shard_sizes)
+
+        # Setup for coordinating.
+        device = torch.device(f'cuda:{world.rank_of_node}')
+        tensor = torch.zeros(1, dtype=torch.int64, device=device)
+
+        # Coordinate the seed across ranks.
+        if world.is_leader:
+            if seed is None:
+                seed = np.random.randint(1 << 60)
+            tensor[0] = seed
+        dist.broadcast(tensor, 0)
+        self.seed = int(tensor)
+
+        # Add a coordinated random prefix to all shm names for uniqueness.
+        if world.is_leader:
+            tensor[0] = np.random.randint(1 << 60)
+        dist.broadcast(tensor, 0)
+        self._prefix = f'{int(tensor):016x}_{self.split}'
+
+        # Set up the epoch counter.
+        #
+        # Note: we do not assume that the end of __iter__() will ever be reached, so we need to
```

Comment:
Can you please explain this sentence `Note: we do not assume that the end of __iter__() will ever be reached, so we need to` more ? I didn't quite get it on why it will never reach end of __iter__()

Reply:
a user might get the number of samples from len(), then step what iter() returns with next() len() number of times